File: app/src/main/java/com/mintthursday/recipe/creation/ingredientcreation/IngredientFragment.java
package com.mintthursday.recipe.creation.ingredientcreation;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;

import com.google.android.material.textfield.MaterialAutoCompleteTextView;
import com.mintthursday.R;
import com.mintthursday.models.Ingredient;

import java.util.Objects;

public class IngredientFragment extends Fragment {

    public static final String ARG_INGREDIENT = "ARG_INGREDIENT";
    public static final String ARG_POSITION = "ARG_POSITION";
    public static final String BUN_INGREDIENT = "BUN_INGREDIENT";
    public static final String BUN_ITEM_POSITION = "BUN_ITEM_POSITION";
    public static final String REQ_INGREDIENT = "REQ_INGREDIENT";

    private Button btnSave;
    private EditText ingrQty;
    private EditText ingrName;
    private MaterialAutoCompleteTextView ingrUnit;
    private final TextWatcher ingrTextWatcher = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {
        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            String ingrInput = ingrName.getText().toString().trim();
            String qtyInput = ingrQty.getText().toString().trim();
            String unitInput = ingrUnit.getText().toString().trim();
            btnSave.setEnabled(!ingrInput.isEmpty() && !qtyInput.isEmpty() && !unitInput.isEmpty());
        }

        @Override
        public void afterTextChanged(Editable s) {

        }
    };
    private int itemPosition = -1;

    public IngredientFragment() {
    }

    public static IngredientFragment newInstance(Ingredient ingredient, int itemPosition) {
        IngredientFragment fragment = new IngredientFragment();
        Bundle bundle = new Bundle();
        bundle.putParcelable(ARG_INGREDIENT, ingredient);
        bundle.putInt(ARG_POSITION, itemPosition);
        fragment.setArguments(bundle);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_ingredient, container, false);
        initToolbar(view);

        btnSave = view.findViewById(R.id.btnSave);
        ingrQty = view.findViewById(R.id.textInputEditTextQty);
        ingrName = view.findViewById(R.id.textInputEditTextName);
        ingrUnit = view.findViewById(R.id.spinnerIngredUnits);

        ingrName.addTextChangedListener(ingrTextWatcher);
        ingrQty.addTextChangedListener(ingrTextWatcher);
        ingrUnit.addTextChangedListener(ingrTextWatcher);

        btnSave.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                saveIngredient();
            }
        });

        String[] type = getResources().getStringArray(R.array.ingredient_units_list);
        ingrUnit.setAdapter(new ArrayAdapter<>(this.getContext(), R.layout.dropdown_menu_ingredient_unit, type));
        if (getArguments() != null && getArguments().getParcelable(ARG_INGREDIENT) != null) {
            initEditIngredient();
        }
        return view;
    }

    private void initToolbar(View view) {
        Toolbar ingredientToolbar = view.findViewById(R.id.toolbar);
        ((AppCompatActivity) getActivity()).setSupportActionBar(ingredientToolbar);
        ingredientToolbar.setNavigationOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ((AppCompatActivity) getActivity()).onBackPressed();
            }
        });
    }

    private void initEditIngredient() {
        Ingredient ingredientEdit = (Ingredient) getArguments().getParcelable(ARG_INGREDIENT);
        itemPosition = getArguments().getInt(ARG_POSITION);
        if (ingredientEdit != null) {
            ingrName.setText(ingredientEdit.getName());
            ingrQty.setText(String.valueOf(ingredientEdit.getQuantity()));
            ingrUnit.setText(ingredientEdit.getUnit());
        }
    }

    private void saveIngredient() {
        String name = String.valueOf(ingrName.getText());
        String stringQuantity = String.valueOf(ingrQty.getText());
        double quantity;
        if (!stringQuantity.isEmpty()) {
            quantity = Double.parseDouble(stringQuantity);
        } else {
            quantity = 0;
        }
        String unit = String.valueOf(ingrUnit.getText());
        Bundle result = new Bundle();
        Ingredient ingredient = new Ingredient(name, quantity, unit);
        result.putParcelable(BUN_INGREDIENT, ingredient);
        result.putInt(BUN_ITEM_POSITION, itemPosition);
        getParentFragmentManager().setFragmentResult(REQ_INGREDIENT, result);
        if (getActivity() != null) {
            getActivity().onBackPressed();
        }
    }
}

File: app/src/main/java/com/mintthursday/recipe/creation/ingredientcreation/IngredientFragment.kt
package com.mintthursday.recipe.creation.ingredientcreation

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import com.mintthursday.App
import com.mintthursday.R
import com.mintthursday.databinding.FragmentIngredientBinding
import com.mintthursday.models.Ingredient

class IngredientFragment : Fragment() {

    private var _binding: FragmentIngredientBinding? = null
    private val binding get() = _binding!!

    private var itemPosition = -1

    private val ingrTextWatcher: TextWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            val ingrInput = binding.ingredientName.text.toString().trim { it <= ' ' }
            val qtyInput = binding.ingredientQty.text.toString().trim { it <= ' ' }
            val unitInput = binding.spinnerIngredientUnits.text.toString().trim { it <= ' ' }
            binding.btnSave.isEnabled = ingrInput.isNotEmpty() && qtyInput.isNotEmpty() && unitInput.isNotEmpty()
        }

        override fun afterTextChanged(s: Editable) {}
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        _binding = FragmentIngredientBinding.inflate(inflater, container, false)

        initToolbar()

        binding.ingredientName.addTextChangedListener(ingrTextWatcher)
        binding.ingredientQty.addTextChangedListener(ingrTextWatcher)
        binding.spinnerIngredientUnits.addTextChangedListener(ingrTextWatcher)

        val type = resources.getStringArray(R.array.ingredient_units_list)
        binding.spinnerIngredientUnits.setAdapter(ArrayAdapter(this.requireContext(), R.layout.dropdown_menu_ingredient_unit, type))

        binding.btnSave.setOnClickListener {
            saveIngredient()
        }

        initEditIngredient()
        return view
    }

    private fun initToolbar() {
        val ingredientToolbar: Toolbar = binding.toolbar
        (activity as? AppCompatActivity)?.setSupportActionBar(ingredientToolbar)
        ingredientToolbar.setNavigationOnClickListener { App.instance.router.exit() }
    }

    private fun initEditIngredient() {
        val arguments = arguments ?: return

        //todo create random Ingredient.id
        itemPosition = arguments.getInt(ARG_POSITION)

        val ingredient = arguments.getParcelable<Ingredient>(ARG_INGREDIENT) ?: return
        binding.ingredientName.setText(ingredient.name)
        binding.ingredientQty.setText(ingredient.quantity.toString())
        binding.spinnerIngredientUnits.setText(ingredient.unit)
    }

    private fun saveIngredient() {
        val ingredient = buildIngredient()

        val bundle = bundleOf(BUN_INGREDIENT to ingredient, BUN_ITEM_POSITION to itemPosition)

        parentFragmentManager.setFragmentResult(REQ_INGREDIENT, bundle)
        App.instance.router.exit()
    }

    private fun buildIngredient(): Ingredient {
        val name = binding.ingredientName.text.toString()
        val stringQuantity = binding.ingredientQty.text.toString()
        val quantity: Double = if (stringQuantity.isNotEmpty()) {
            stringQuantity.toDouble()
        } else {
            0.0
        }

        val unit = binding.spinnerIngredientUnits.text.toString()

        return Ingredient(name, quantity, unit)
    }

    companion object {

        const val ARG_INGREDIENT = "ARG_INGREDIENT"
        const val ARG_POSITION = "ARG_POSITION"
        const val BUN_INGREDIENT = "BUN_INGREDIENT"
        const val BUN_ITEM_POSITION = "BUN_ITEM_POSITION"
        const val REQ_INGREDIENT = "REQ_INGREDIENT"

        fun newInstance(ingredient: Ingredient?, itemPosition: Int): IngredientFragment {
            val fragment = IngredientFragment()
            fragment.arguments = bundleOf(ARG_INGREDIENT to ingredient, ARG_POSITION to itemPosition)
            return fragment
        }
    }
}